File: 2025-2/sesion-de-clase-11/SoftProgNegocio/BOImpl/OrdenVentaBOImpl.cs
using System;
using System.Collections.Generic;
using System.Data.Common;
using PUCP.SoftProg.Db;
using PUCP.SoftProg.Modelo;
using PUCP.SoftProg.Modelo.Ventas;
using PUCP.SoftProg.Negocio.BO;
using PUCP.SoftProg.Persistencia.DAO.Ventas;
using PUCP.SoftProg.Persistencia.DAOImpl.Ventas;

namespace PUCP.SoftProg.Negocio.BOImpl {
    public class OrdenVentaBOImpl : IOrdenVentaBO {
        private readonly IOrdenVentaDAO ordenVentaDAO;
        private readonly ILineaOrdenVentaDAO lineaOrdenVentaDAO;

        public OrdenVentaBOImpl() {
            this.ordenVentaDAO = new OrdenVentaDAOImpl();
            this.lineaOrdenVentaDAO = new LineaOrdenVentaDAOImpl();
        }

        public void Guardar(OrdenVenta ordenVenta, Estado estado) {
            using (DbConnection conexion = DBFactoryProvider.GetManager().GetConnection()) {
                conexion.Open();
                DbTransaction transaction = conexion.BeginTransaction();
                try {
                    switch (estado) {
                        case Estado.Nuevo:
                            int idOrden = this.ordenVentaDAO.Crear(ordenVenta, transaction);
                            ordenVenta.Id = idOrden;

                            foreach (LineaOrdenVenta linea in ordenVenta.LineasOrdenVenta) {
                                linea.OrdenVenta = new OrdenVenta { Id = idOrden };
                                this.lineaOrdenVentaDAO.Crear(linea, transaction);
                            }
                            break;

                        case Estado.Modificado:
                            this.ordenVentaDAO.Actualizar(ordenVenta, transaction);

                            foreach (LineaOrdenVenta linea in ordenVenta.LineasOrdenVenta) {
                                if (linea.Id == 0) {
                                    linea.OrdenVenta = new OrdenVenta { Id = ordenVenta.Id };
                                    this.lineaOrdenVentaDAO.Crear(linea, transaction);
                                }
                                else {
                                    this.lineaOrdenVentaDAO.Actualizar(linea, transaction);
                                }
                            }
                            break;
                    }

                    transaction.Commit();
                }
                catch (Exception ex) {
                    Console.Error.WriteLine("Error al guardar la OrdenVenta: " + ex.Message);
                    transaction.Rollback();
                    throw;
                }
            }
        }

        public void Eliminar(int id) {
            using (DbConnection conexion = DBFactoryProvider.GetManager().GetConnection()) {
                conexion.Open();
                DbTransaction transaction = conexion.BeginTransaction();
                try {
                    List<LineaOrdenVenta> lineas = this.lineaOrdenVentaDAO.LeerTodosPorOrden(id, transaction);
                    foreach (LineaOrdenVenta linea in lineas) {
                        this.lineaOrdenVentaDAO.Eliminar(linea.Id, transaction);
                    }

                    if (!this.ordenVentaDAO.Eliminar(id, transaction)) {
                        throw new Exception("La OrdenVenta no se pudo eliminar.");
                    }

                    transaction.Commit();
                }
                catch (Exception ex) {
                    Console.Error.WriteLine("Error al eliminar la OrdenVenta: " + ex.Message);
                    transaction.Rollback();
                    throw;
                }
            }
        }

        public OrdenVenta Obtener(int id) {
            return this.ordenVentaDAO.Leer(id);
        }

        public List<OrdenVenta> Listar() {
            return this.ordenVentaDAO.LeerTodos();
        }
    }
}

File: 2025-2/sesion-de-clase-19/java/SoftProg/SoftProgNegocio/src/main/java/pe/edu/pucp/inf30/softprog/boimpl/ventas/OrdenVentaBOImpl.java
package pe.edu.pucp.inf30.softprog.boimpl.ventas;

import java.sql.Connection;
import java.sql.SQLException;
import java.util.List;
import pe.edu.pucp.inf30.softprog.bo.ventas.OrdenVentaBO;
import pe.edu.pucp.inf30.softprog.dao.ventas.LineaOrdenVentaDAO;
import pe.edu.pucp.inf30.softprog.dao.ventas.OrdenVentaDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.ventas.LineaOrdenVentaDAOImpl;
import pe.edu.pucp.inf30.softprog.daoimpl.ventas.OrdenVentaDAOImpl;
import pe.edu.pucp.inf30.softprog.db.DBFactoryProvider;
import pe.edu.pucp.inf30.softprog.db.DBManager;
import pe.edu.pucp.inf30.softprog.modelo.Estado;
import pe.edu.pucp.inf30.softprog.modelo.ventas.OrdenVenta;
import pe.edu.pucp.inf30.softprog.modelo.ventas.LineaOrdenVenta;

/**
 *
 * @author eric
 */
public class OrdenVentaBOImpl implements OrdenVentaBO {
    private final OrdenVentaDAO ordenVentaDao;
    private final LineaOrdenVentaDAO lineaOrdenVentaDao;
    
    public OrdenVentaBOImpl() {
        this.ordenVentaDao = new OrdenVentaDAOImpl();
        this.lineaOrdenVentaDao = new LineaOrdenVentaDAOImpl();
    }
    
    @Override
    public List<OrdenVenta> listar() {
        return this.ordenVentaDao.leerTodos();
    }

    @Override
    public OrdenVenta obtener(int id) {
        OrdenVenta orden = ordenVentaDao.leer(id);
        if (orden == null) return null;
        
        List<LineaOrdenVenta> lineas = 
                lineaOrdenVentaDao.leerTodosPorOrden(id);
        orden.setLineas(lineas);
        return orden;
    }

    @Override
    public void eliminar(int id) {
        DBManager dbManager = DBFactoryProvider.getManager();
        try (Connection conn = dbManager.getConnection()) {
            conn.setAutoCommit(false);

            try {
                List<LineaOrdenVenta> lineas = 
                        lineaOrdenVentaDao.leerTodosPorOrden(id, conn);
                for (LineaOrdenVenta linea : lineas) {
                    if (linea.getOrdenVenta().getId() == id) {
                        lineaOrdenVentaDao.eliminar(linea.getId(), conn);
                    }
                }

                if (!ordenVentaDao.eliminar(id, conn)) {
                    throw new RuntimeException("La Orden: " + id + ", "
                            + "no se pudo eliminar");
                }
                conn.commit();
            } catch (SQLException ex) {
                conn.rollback();
                throw new RuntimeException("Error eliminando OrdenVenta "
                        + "con id=" + id, ex);
            }
        } catch (ClassNotFoundException | SQLException e) {
            throw new RuntimeException("Error de conexión al eliminar "
                    + "OrdenVenta", e);
        }
    }

    @Override
    public void guardar(OrdenVenta modelo, Estado estado) {
        DBManager dbManager = DBFactoryProvider.getManager();
        try (Connection conn = dbManager.getConnection()) {
            conn.setAutoCommit(false);

            try {
                switch (estado) {
                    case Nuevo -> {
                        int idOrden = this.ordenVentaDao.crear(modelo, conn);
                        modelo.setId(idOrden);
                        for (LineaOrdenVenta linea : modelo.getLineas()) {
                            linea.setOrdenVenta(modelo);
                            lineaOrdenVentaDao.crear(linea, conn);
                        }
                    }
                    case Modificado -> {
                        ordenVentaDao.actualizar(modelo, conn);
                        for (LineaOrdenVenta linea : modelo.getLineas()) {
                            if (linea.getId() == 0) {
                                linea.setOrdenVenta(modelo);
                                lineaOrdenVentaDao.crear(linea, conn);
                            } else {
                                lineaOrdenVentaDao.actualizar(linea, conn);
                            }
                        }
                    }
                }

                conn.commit();
            } catch (SQLException ex) {
                conn.rollback();
                throw new RuntimeException("Error guardando OrdenVenta", ex);
            }
        } catch (ClassNotFoundException | SQLException e) {
            throw new RuntimeException("Error de conexión al guardar OrdenVenta"
                    + "", e);
        }
    }

    @Override
    public List<OrdenVenta> listarOrdenesVentaPorCuenta(String cuenta) {
        return this.ordenVentaDao.listarOrdenesVentaPorCuenta(cuenta);
    }
}
